### PR TITLE
(Hopefully) final revisions for lesson import

### PR DESCRIPTION
**Pull request by bethanyaconnor**
- Instead of setting the `slide` setting on activity sections, replace the CurriculumBuilder syntax with `[slide]`
- Handle tip icons (tip links) differently to handle the above change as well as various other instances of tip icons being inline
    - Preprocess the activity content for everything we can assume starts at the beginning of a line, such as headings, remark blocks, tips, and code studio pullthrough. Notably, tip icons are not in this list.
    - After getting all the stuff from the above step, there are blocks of markdown that have zero or more tip icons embedded in them. A new function, `create_activity_sections_from_markdown` looks for tip icons, creating activity sections along the way.
    - These changes required another fundamental change in how we look for tip icons, which is to look anywhere within a line, strip the tip icon syntax out of the markdown and set a tip on that activity section.
    - Note that remarks may also have tips within them. This largely "just worked" because we were already looking for tips within remarks

Stuff that still is not supported in this import:
- Tips in headers
- Multiple tips on the same activity section. This is supported in our system but I didn't get a change to implement this. There are fairly few instances of this so hopefully this won't be too arduous to fix up.

## Links

- [spec]()
- [jira]()

## Testing story

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: lib/cdo/lesson_import_helper.rb [hunk 1]
```diff
@@ -411,6 +387,7 @@ def self.find_markdown_chunks(markdown, existing_matches)
       matches.push({index: 0, type: 'markdown', substring: substring}) unless substring.empty?
     end
     (0...existing_matches.length).each do |i|
+      #next if !existing_matches[i][:substring].empty? && i > 0 && existing_matches[i-1][:substring].include?(existing_matches[i][:substring])
```
Comment (dmcavoy) [inline lib/cdo/lesson_import_helper.rb]: still needed?

File: lib/cdo/lesson_import_helper.rb [hunk 2]
```diff
@@ -102,51 +96,45 @@ def self.create_activity_sections(activity_markdown, lesson_activity_id, levels)
 
     sections = []
     name = ''
-    position = 1
```
Comment (dmcavoy) [inline lib/cdo/lesson_import_helper.rb:105]: I noticed position was removed. Don't need it anymore?

File: lib/cdo/lesson_import_helper.rb (hunk 1 above)
Comment (bethanyaconnor) [reply lib/cdo/lesson_import_helper.rb]: Thanks for catching! Removed.

File: lib/cdo/lesson_import_helper.rb (hunk 2 above)
Comment (bethanyaconnor) [reply lib/cdo/lesson_import_helper.rb:105]: They're regenerated down on line 150 so I decided this wasn't serving a purpose anymore. Activity sections need a position in order to be saved so I changed it to always set the position to 0 on line 128. I'm happy to switch back here though!

File: lib/cdo/lesson_import_helper.rb [hunk 3]
```diff
@@ -75,19 +75,13 @@ def self.create_lesson_resources(cb_resources, course_version_id)
   end
 
   def self.create_activity_sections(activity_markdown, lesson_activity_id, levels)
-    # Find any special syntax, such as tips or remarks, and gather them.
-    # TODO tips and remarks both show up in tip_matches. We filter out remarks
-    # but we should try to do something a bit smarter here.
+    activity_markdown.gsub!(/slide!!!slide-\d+(?:<!-- place where you'd like the icon -->)?/, "[slide]")
```
Comment (bethanyaconnor) [inline lib/cdo/lesson_import_helper.rb:78]: In this line and the one below, the markdown is searched for "easy" replacements. This line looks for the old slide syntax and replaces it with our new `[slide]` syntax. The line below looks for `[guide]` syntax, which we're not using anymore, and removes it.

I don't know why I didn't do this for `[guide]` before now to be honest but oh well.

File: lib/cdo/lesson_import_helper.rb [hunk 4]
```diff
@@ -75,19 +75,13 @@ def self.create_lesson_resources(cb_resources, course_version_id)
...
+    activity_markdown.gsub!(/\[\/?guide\]/, '')
     tip_matches = find_tips(activity_markdown).select {|m| m[1] != 'say'}.map {|m| {index: activity_markdown.index(m[0]), type: 'tip', match: m, substring: m[0]}}
```
Comment (bethanyaconnor) [inline lib/cdo/lesson_import_helper.rb:80]: I alluded to this in the description but tips, names, pullthrough, and remarks are all syntax that _must_ be at the beginning of a line. This means that a line cannot both be a pullthrough and a remark, for instance. This allows us to look for all of these at the same time and process them all at the same time.

File: lib/cdo/lesson_import_helper.rb [hunk 5]
```diff
@@ -102,51 +96,45 @@ def self.create_activity_sections(activity_markdown, lesson_activity_id, levels)
...
         # add the level activity sections here.
         if pullthrough_match[1].blank? || ([1, 2].include?(pullthrough_match[1]) && levels.length == pullthrough_match[2].to_i)
-          level_sections = create_activity_sections_by_progression(levels, lesson_activity_id, position)
+          level_sections = create_activity_sections_by_progression(levels, lesson_activity_id)
           sections += level_sections
           levels.clear
-          position += level_sections.length
         end
         next
+      elsif match[:type] == 'remark'
```
Comment (bethanyaconnor) [inline lib/cdo/lesson_import_helper.rb:121]: FWIW, moving this down here was a mistake (I removed it then added it back). Sorry for adding more noise to this already hard to read diff 😞

File: lib/cdo/lesson_import_helper.rb [hunk 6]
```diff
@@ -102,51 +96,45 @@ def self.create_activity_sections(activity_markdown, lesson_activity_id, levels)
...
         activity_section = ActivitySection.new
         key = match[:match][3]&.strip || "#{match[:match][1]}-0"
         activity_section.tips = [create_tip(key, match[:match][1] || "tip", match[:match][4])]
         activity_section.key ||= SecureRandom.uuid
+        activity_sections = [activity_section]
         match[:activity_section_key] = activity_section.key
       elsif match[:type] == 'name'
         name = match[:match][1]
         next
-      elsif match[:type] == 'tiplink'
```
Comment (bethanyaconnor) [inline lib/cdo/lesson_import_helper.rb:119]: Tip icons (tip links), however, might not be at the beginning of a line so we're going to handle them differently. This is the biggest change introduced in this PR, but it allows us to parse lines like:
1. `* tip!!!tip-0 my bulleted point`
1. `1. tip!!!tip-0 my numbered point`
1. `[slide] tip!!!tip-0 my sentence that has both a slide icon and a tip icon`

The third of these was introduced by the fact that we changed the slide icon syntax, the other big change in this PR. We could have shoehorned something in with the old logic but changing the logic gave us better parsing for the other two examples for free.

File: lib/cdo/lesson_import_helper.rb [hunk 7]
```diff
@@ -327,14 +310,14 @@ def self.find_tips(markdown)
...
-  def self.find_tip_links(markdown)
+  def self.find_first_tip_link(markdown)
     # See https://github.com/code-dot-org/remark-plugins/blob/master/src/tiplink.js
     # Looks for the location where tip icons should appear
     # Example: tip!!!tip-0<!-- place where you'd like the icon --> some text
     # <!-- place where you'd like the icon --> is optional but is written out
     # in this regex in order to be able to correctly the text that should be displayed
-    regex = /^([\w-]+)!!! ?([\w-]+)(?:<!-- place where you'd like the icon -->)?(.*\n?.*)$/
-    markdown.to_enum(:scan, regex).map {Regexp.last_match}
+    regex = /^?(\S+ +)?([\w-]+)!!! ?([\w-]+)(?:<!-- place where you'd like the icon -->)?(.*\n?.*)$/
```
Comment (bethanyaconnor) [inline lib/cdo/lesson_import_helper.rb:319]: This change allows (optionally) for some non-whitespace characters followed by at least one space to come before tip icons.

File: lib/cdo/lesson_import_helper.rb [hunk 8]
```diff
@@ -353,39 +336,17 @@ def self.create_tip(type, key, markdown)
...
+    description = (tip_link_match[1] || '') + tip_link_match[4]
     unless tip
-      return ActivitySection.new(description: tip_link_match[3].strip)
+      return ActivitySection.new(description: description)
     end
     tip[:paired] = true
     tip_match = tip[:match]
     activity_section = ActivitySection.new
 
-    # Sometimes theres the the slide icon here as well. Check for that and apply if needed.
```
Comment (bethanyaconnor) [inline lib/cdo/lesson_import_helper.rb:380]: A lot of this removed code is just getting rid of the slides 😁

File: lib/cdo/lesson_import_helper.rb [hunk 9]
```diff
@@ -396,11 +357,26 @@ def self.find_activity_section_names(markdown)
...
-  def self.create_basic_activity_section(markdown)
+  # We preprocess activity markdown to look for remarks, pullthrough, names, and tips
+  # but we need to look for tips (which may not be at the beginning of a line) and
+  # create pure markdown sections.
+  def self.create_activity_sections_from_markdown(markdown, tip_match_map)
+    sections = []
     stripped_markdown = markdown.strip
-    description = stripped_markdown
-    ActivitySection.new(description: description)
+    tip = find_first_tip_link(stripped_markdown)
```
Comment (bethanyaconnor) [inline lib/cdo/lesson_import_helper.rb:366]: This is the implementation of the new tip logic. It looks for new tips, creating activity sections as it goes.

For [example](https://www.codecurricula.com/csp-21/unit4/9/):
```**Distribute:** Share print copies of [r song-lyrics-functions] or project the activity guide where all can see it.

[slide] discussion!!!discussion-0 **Prompt:**

**Discuss:** Have students silently consider these prompts on their own, then share with a neighbor, and finally move to a full class discussion. 
```

The first tip is at `[slide] discussion!!!discussion-0 ***Prompt:***` so we create an activity section with everything above in on line 370, then create another activity section with the tip on line 373. Then there are no more tips found, so the while loops breaks and we create an activity section with the rest of the content on line 378.

So the descriptions of the 3 activity sections are:
1. `*Distribute:** Share print copies of [r song-lyrics-functions] or project the activity guide where all can see it.`
2. `[slide] **Prompt:**` (the tip syntax is stripped out)
3. `**Discuss:** Have students silently consider these prompts on their own, then share with a neighbor, and finally move to a full class discussion.`

**Comment by bethanyaconnor** [comment]
Sorry for the spam! I've tried to walk through all the changes but let me know if you have any questions!

**Review by dmcavoy — APPROVED** [review]
Thanks for all the comments Bethany. This is working great on the adhoc and looks good in the code so lets merge it

**Pull request merged**